packages/common/src/EthersTransactionUtils.ts

```diff
@@ -0,0 +1,241 @@
+
+// Every Error should be a Record, so this type guard helps to check on individual properties of the Error object.
+function isErrorRecord(error: unknown): error is Error & Record<PropertyKey, unknown> {
+  return error instanceof Error;
+}
+
+// Type guard for known Ethers error codes.
+function isEthersErrorCode(code: unknown): code is ErrorCode {
+  return typeof code === "string" && Object.values(ErrorCode).includes(code as ErrorCode);
+}
+
+// Every Ethers error should be a Record, so this type guard helps to check on individual properties of the Ethers error.
+function isEthersError(error: unknown): error is EthersV5Error & Record<PropertyKey, unknown> {
+  return isErrorRecord(error) && isEthersErrorCode(error.code);
+}
+
+// Type guard for potential Ethers error to get revert data when gas estimation fails.
+function isEthersEstimateGasError(error: unknown): error is EthersV5EstimateGasError {
+  return (
+    isEthersError(error) &&
+    error.code === ErrorCode.UNPREDICTABLE_GAS_LIMIT &&
+    isRecordStringUnknown(error.error) &&
+    isRecordStringUnknown(error.error.error) &&
+    ethersUtils.isHexString(error.error.error.data)
+  );
+}
+
+// Type guard that can be used by the caller to handle revert data in the Ethers transaction runner errors.
+export function isEthersTxRunnerError(error: unknown): error is EthersTxRunnerError {
+  return isErrorRecord(error) && ethersUtils.isHexString(error.revertData);
+}
+
+// Type guard for transaction receipt thrown when sending a transaction. This only checks the blockNumber that is needed
+// for the replay when extracting the revert data.
+function isEthersTransactionReceipt(receipt: unknown): receipt is TxReceiptWithBlockNumber {
+  return isRecordStringUnknown(receipt) && typeof receipt.blockNumber === "number";
+}
+
+// Type guard for Ethers wait sent transaction error. This checks the error code, the transaction and receipt structures
+// that are needed for the replay when extracting the revert data.
+function isEthersWaitTransactionError(error: unknown): error is EthersV5WaitTransactionError {
+  return isEthersError(error) && error.code === ErrorCode.CALL_EXCEPTION && isEthersTransactionReceipt(error.receipt);
+}
+
+// Helper function to add revert data to thrown error in the Ethers transaction wait() method. It overrides the original
+// wait() method and adds the logic to replay the transaction to extract the revert data if the transaction fails.
+function overrideTxWait<T extends TransactionResponse | ContractTransaction>(tx: T, signer: Signer): T {
+  const originalWait = tx.wait.bind(tx);
+  tx.wait = async (confirmations?: number): Promise<WaitReturn<T>> => {
+    try {
+      // Call the original wait() method to get the receipt.
+      return await originalWait(confirmations);
+    } catch (error) {
+      // The transaction could still have failed when mined due to changed state after the gas estimation.
+      if (isEthersWaitTransactionError(error)) {
```

Comment:
Do we always need to override the `tx.wait()` or are there cases where the original `tx.wait()` error may have:
```
    e?.data ??
    e?.error?.data ??
    e?.error?.error?.data
```

Reply:
We are trying to catch [this](https://github.com/ethers-io/ethers.js/blob/ec1b9583039a14a0e0fa15d0a2a6082a2f41cf5b/packages/providers/src.ts/base-provider.ts#L1548-L1554) error when `receipt.status === 0`. There might be also other errors, e.g. tx got replaced when waiting for the receipt, but those would not have revert data available.